src/algorithm/kinematics.hpp

```diff
@@ -148,5 +148,36 @@ namespace pinocchio
 /* --- Details -------------------------------------------------------------------- */
 #include "pinocchio/algorithm/kinematics.hxx"
 
+#if PINOCCHIO_ENABLE_TEMPLATE_INSTANTIATION
```

Comment:
To not pollute too much the declaration file, I would suggest putting this explicit inside another file named something like: `kinematics.txx`

Reply:
As you want. You can also put it in the `.hxx` file.

Reply:
Yep, this is a solution. But in fact, I would like to have the opportunity to perform the template instantiation for various scalar types ... and to be able to use them in the same compilation units, for instance, mixing `double` and `CppAD::AD<double>`. For that, I think we need to use similar tricks than the one presents in boost PP. Maybe, @jmirabel you might have some clues here.

Reply:
For the definition part, I think you can use CMake to achieve this. Let the `cc` file be
```cpp
#include ...

// version 1
template class pinocchio::DataTpl<SCALAR>;
// or version 2
// template class pinocchio::DataTpl<@SCALAR_TYPE@>;
```
then, in CMake, it depends if you want to put them in separated libs:
- separate libs: you only need to add a private compilation definition (version 1)
- one single lib: use cmake to generate cc files from version 2.

For the header part, options I can think of are:
- Boost PP ( a bit tedious I think)
- CMake generated files using a similar trick as above

Reply:
Very nice for the CMake shortcut. We may think something very generic.

Reply:
For the lib, I think to have a separate lib which is then linked to the main one would be the simplest and most efficient way.

Reply:
On this topic, I think I cleared up my mind a bit. Here is what, I think, we need.

- add CMake template header and definition files, that allows to create template instantiation with a given *scalar* and *JointModelCollectionTpl* (this is needed for my use-case :) ).
- add a CMake macro that allows to configure the header and definition file.
- install the macro and the template files so that dependent package can generate their own template instantiation files.

Reply:
OK. I do agree. So we also need the Options (for alignment issues and so on) to be complete.
To simplify the extension process, for each module (files in the algorithm section), we need just to define a .txx with all the generated templates which assume defined:
- scalar_type, options_values, JointModelCollection_type
And these 3 values will be generated by CMake, and for each context, will include the .txx file.

Reply:
For the inclusion, there are two cases:
- for the default(s) static ones (if any) provided by the library, adding the correct `#include <....txx>` in *....hpp*.
- for the one generated elsewhere, I think we can let the user include them.